### ISSUE TITLE
Result was devaites from the value in paper. 

### ISSUE THREAD

**Issue by Shylock-H** [bkkgbkjb/OPPO#1]
I am facing a performance issue while running this project. I followed parameter settings in your paper and run the experiment on 'hopper-medium-replay-v2' dataset and it got the final normalize score about 17, which is deviates significantly from the reported 88.9 in the paper. Here is my result:

![image](https://github.com/bkkgbkjb/OPPO/assets/70041484/88701fc3-b653-4f84-855b-b56f2e980ff8)

Thank you for your attention to this matter. I appreciate your help.

**Comment by bkkgbkjb** [comment]
Thanks for the report

May I ask specifically which experiment you're refering to?

Human or scripted teacher?
Docker or manual installation?
and which python command you're using?

**Comment by Shylock-H** [comment]
I run the python file in './oppo/human/gym/experiment.py' with manual installation. And my python command is 'python experiment.py --env hopper --dataset medium-replay --K 20'. Here is my config file.
{
    "K": 20,
    "activation_function": "relu",
    "batch_size": 64,
    "dataset": "medium-replay",
    "device": "cuda",
    "dropout": 0.1,
    "embed_dim": 128,
    "env": "hopper",
    "eval_no_change": true,
    "force_save_model": false,
    "foresee": 200,
    "learning_rate": 0.0001,
    "log_to_wandb": true,
    "max_iters": 100,
    "mode": "normal",
    "model_type": "pdt",
    "n_head": 2,
    "n_layer": 3,
    "num_eval_episodes": 50,
    "num_steps_per_iter": 1000,
    "pct_traj": 1.0,
    "phi_norm_loss_ratio": 0.1,
    "pref_loss_ratio": 10,
    "robosuite_dataset_path": "./robosuite_dataset",
    "seed": 0,
    "subepisode": true,
    "train_no_change": true,
    "w_lr": 0.01,
    "warmup_steps": 10000,
    "weight_decay": 0.0001,
    "z_dim": 16
}

**Comment by bkkgbkjb** [comment]
Dear Shylock

after some investigations, we do find 2 bugs that may lead to the bad performance on your side:
1. evaluation `max_time_step` is hard-coded to `500`, which would make inconsitence on `mujoco` envs
2. a hyperparameters `K`, length of trajectories, should be changed to `100` according to our logbook
3. seeds used in our side are mainly `3333 4444 5555 6666`, etc.

Reproduction experiments we run on our side show  performance constistent with our reports in paper
and we attach our code diff and exps results in the following

**Comment by bkkgbkjb** [comment]
This is result of running `git diff` in our latest code, we would also update main code base later


```
diff --git a/oppo/human/gym/args.py b/oppo/human/gym/args.py
index a884718..c372a39 100644
--- a/oppo/human/gym/args.py
+++ b/oppo/human/gym/args.py
@@ -1,8 +1,8 @@
 import argparse
 
 parser = argparse.ArgumentParser()
-parser.add_argument('--env', type=str, default='Lift')
-parser.add_argument('--dataset', type=str, default='mh')  # medium, medium-replay, medium-expert, expert
+parser.add_argument('--env', type=str, default='hopper')
+parser.add_argument('--dataset', type=str, default='medium-expert')  # medium, medium-replay, medium-expert, expert
 parser.add_argument('--mode', type=str, default='normal')  # normal for standard setting, delayed for sparse
 parser.add_argument('--K', type=int, default=100)
 parser.add_argument('--pct_traj', type=float, default=1.)
@@ -16,14 +16,13 @@ parser.add_argument('--dropout', type=float, default=0.1)
 parser.add_argument('--learning_rate', '-lr', type=float, default=1e-4)
 parser.add_argument('--weight_decay', '-wd', type=float, default=1e-4)
 parser.add_argument('--warmup_steps', type=int, default=10000)
-parser.add_argument('--num_eval_episodes', type=int, default=50)
-parser.add_argument('--seed', type=int, default=0)
+parser.add_argument('--num_eval_episodes', type=int, default=10)
+parser.add_argument('--seed', type=int, default=3333)
 parser.add_argument('--max_iters', type=int, default=100)
 parser.add_argument('--z-dim', type=int, default=16)
 parser.add_argument('--foresee', type=int, default=200)
 parser.add_argument('--num_steps_per_iter', type=int, default=1000)
 parser.add_argument('--device', type=str, default='cuda')
-parser.add_argument('--robosuite_dataset_path', type=str, default='./robosuite_dataset')
 parser.add_argument('--train-no-change', type=bool, default=True)
 parser.add_argument('--eval-no-change', type=bool, default=True)
 parser.add_argument('--subepisode', type=bool, default=True)
diff --git a/oppo/human/gym/experiment.py b/oppo/human/gym/experiment.py
index 3b66718..5afc806 100644
--- a/oppo/human/gym/experiment.py
+++ b/oppo/human/gym/experiment.py
@@ -25,9 +25,6 @@ from reporter import get_reporter
 
 from d4rl.infos import REF_MIN_SCORE, REF_MAX_SCORE
 import os
-import d4rl
-from utils import make_robosuite_env_and_dataset
-from tqdm import tqdm
 
 
 def discount_cumsum(x, gamma):
@@ -51,7 +48,6 @@ def experiment(
     model_type = variant['model_type']
     group_name = f'{exp_prefix}-{env_name}-{dataset}'
     exp_prefix = f'{group_name}-{random.randint(int(1e5), int(1e6) - 1)}'
-    dataset_path = os.path.join(variant['robosuite_dataset_path'], variant['env'].lower(), variant['dataset'], "low_dim.hdf5")
 
     if env_name == 'hopper':
         env = gym.make('Hopper-v3')
@@ -74,23 +70,6 @@ def experiment(
         max_ep_len = 100
         env_targets = [76, 40]
         scale = 10.
-    elif env_name =='antmaze-medium':
-        env = gym.make(f'antmaze-medium-{dataset}-v2')
-        max_ep_len=1001
-        scale=1.
-        env_targets=[1.]
-    elif env_name == 'antmaze-large':
-        env = gym.make(f'antmaze-large-{dataset}-v2')
-        max_ep_len=1001
-        env_targets=[1.]
-        scale=1.
-    elif env_name in ['Lift', 'Can']:
-        env, _ = make_robosuite_env_and_dataset(env_name, variant['seed'], dataset_path, max_episode_steps=500)
-        max_ep_len = 500
-        if env_name=='Can' and dataset == 'mh':
-            max_ep_len = 1050
-        env_targets = [1.]
-        scale=1.
     else:
         raise NotImplementedError
     env.seed(variant['seed'])
@@ -103,7 +82,7 @@ def experiment(
 
     # load dataset
     dir_path = variant.get('dirpath', '.')
-    dataset_path = f'{dir_path}/data/{env_name}_{dataset}.pkl'
+    dataset_path = f'{dir_path}/data/{env_name}-{dataset}-v2.pkl'
     with open(dataset_path, 'rb') as f:
         trajectories = pickle.load(f)
 
@@ -143,8 +122,8 @@ def experiment(
     print(f'z_dim is: {z_dim}')
     print(f"reward foresee is: {variant['foresee']}")
 
-    expert_score = 1.
-    random_score = 0.
+    expert_score = REF_MAX_SCORE[f"{variant['env']}-{variant['dataset']}-v2"]
+    random_score = REF_MIN_SCORE[f"{variant['env']}-{variant['dataset']}-v2"]
     print(f"max score is: {expert_score}, min score is {random_score}")
 
     # only train on top pct_traj trajectories (for %BC experiment)
@@ -170,7 +149,7 @@ def experiment(
             ),
             "actions": torch.tensor(t["actions"], dtype=torch.float32, device=device),
             "rewards": torch.tensor(t["rewards"], dtype=torch.float32, device=device),
-            # "terminals": torch.tensor(t["terminals"], dtype=torch.bool, device=device),
+            "terminals": torch.tensor(t["terminals"], dtype=torch.bool, device=device),
         }
         for t in trajectories
     ]
@@ -251,8 +230,8 @@ def experiment(
 
         return s, a, None, None, None, timesteps, mask
 
-    # deal with human_labeled data
-    datadir = f'./data/dataset/human_label/{env_name}_{dataset}/data.pkl'
+    # 处理human_label的数据
+    datadir = f'./data/dataset/human_label/{env_name}-{dataset}-v2/data.pkl'
     with open(datadir, 'rb') as f:
         ds = pickle.load(f)
     s1 = torch.as_tensor(ds['observations'], dtype=torch.float32, device=device)
@@ -329,7 +308,7 @@ def experiment(
         def fn(model):
             nonlocal largest_norm_return_mean
             returns, norm_returns, lengths = [], [], []
-            for _ in tqdm(range(num_eval_episodes)):
+            for _ in range(num_eval_episodes):
                 with torch.no_grad():
                     if model_type == 'dt':
                         ret, length = evaluate_episode_rtg(
@@ -337,7 +316,7 @@ def experiment(
                             state_dim,
                             act_dim,
                             model,
-                            max_ep_len=500,
+                            max_ep_len=max_ep_len,
                             scale=scale,
                             target_return=target_rew/scale if not variant['subepisode'] else target_rew / scale,
                             mode=mode,
@@ -352,7 +331,7 @@ def experiment(
                             state_dim,
                             act_dim,
                             model[0],
-                            max_ep_len=500,
+                            max_ep_len=max_ep_len,
                             scale=scale,
                             phi=(model[1]).unsqueeze(0),
                             # phi=(model[1] / torch.linalg.vector_norm(model[1])).unsqueeze(0),
@@ -367,7 +346,7 @@ def experiment(
                             state_dim,
                             act_dim,
                             model,
-                            max_ep_len=500,
+                            max_ep_len=max_ep_len,
                             target_return=target_rew/scale,
                             mode=mode,
                             state_mean=state_mean,
@@ -538,8 +517,7 @@ def experiment(
     for iter in range(variant['max_iters']):
         outputs = trainer.train_iteration(num_steps=variant['num_steps_per_iter'], iter_num=iter+1, print_logs=True)
         norm_return_mean = outputs[f'evaluation/target_{env_targets[0]}_norm_return_mean']
-        # if (variant['force_save_model'] or variant.get("in_tune", False)) and iter >= 30 and iter % 2 == 0:
-        if (variant['force_save_model'] or variant.get("in_tune", False)) and iter >= 20 and (norm_return_mean > max_perf or (iter - last_saved_idx >= 20)):
+        if (variant['force_save_model'] or variant.get("in_tune", False)) and iter >= 10 and (norm_return_mean > max_perf or (iter - last_saved_idx > 20)):
             folder = f"./model_weight_{name}"
             if not os.path.exists(folder):
                 os.mkdir(folder)

```

**Comment by bkkgbkjb** [comment]
This is our runtime arguments, as saved in `runs/.../desc.json`

```json
{
    "K": 100,
    "activation_function": "relu",
    "batch_size": 64,
    "dataset": "medium-replay",
    "device": "cuda",
    "dropout": 0.1,
    "embed_dim": 128,
    "env": "hopper",
    "eval_no_change": true,
    "force_save_model": false,
    "foresee": 200,
    "learning_rate": 0.0001,
    "log_to_wandb": true,
    "max_iters": 100,
    "mode": "normal",
    "model_type": "pdt",
    "n_head": 2,
    "n_layer": 3,
    "num_eval_episodes": 10,
    "num_steps_per_iter": 1000,
    "pct_traj": 1.0,
    "phi_norm_loss_ratio": 0.1,
    "pref_loss_ratio": 10,
    "seed": 3333,
    "subepisode": true,
    "train_no_change": true,
    "w_lr": 0.01,
    "warmup_steps": 10000,
    "weight_decay": 0.0001,
    "z_dim": 16
}
```

**Comment by bkkgbkjb** [comment]
This is our `tensorboard` logged performance

note we run these experiments using `docker` for scalability
but manual installation shall have similar results

seed == 3333:
![image](https://github.com/bkkgbkjb/OPPO/assets/14927934/83e539f0-9174-405c-891b-16bbd6a0f6a0)

seed == 4444:
![image](https://github.com/bkkgbkjb/OPPO/assets/14927934/e03b69a8-d191-4255-8da7-2e786c1f565d)

seed == 5555:
![image](https://github.com/bkkgbkjb/OPPO/assets/14927934/4dc4c228-13a2-4e8f-b287-6cc826960138)

seed == 6666:
![image](https://github.com/bkkgbkjb/OPPO/assets/14927934/37b51c80-8bc3-4e70-bc9d-f1eef32a5405)

**Comment by Shylock-H** [comment]
And here is my result following the same parameters setttings except a random seed of 0.
![image](https://github.com/bkkgbkjb/OPPO/assets/70041484/aae77ea4-d8ad-4ff2-9839-1abeab920f05)

**Comment by bkkgbkjb** [comment]
Yes

Basically, there are several reasons we think for this unconvergent performance:
1. Algorithm itselfs mimics a trade-off between `imitation learning` and `optimal searching/extrapolating` (in the axis of `z`)
2. human labels may contain noise or inconsistent
3. our exps. in `human-mujoco` env are mostly conducted during rebuttal section and have not too much time for hyperparameter tuning


We of course appreciate thoughts/ideas in improving the stability/convergence of `OPPO`
